### ISSUE TITLE
pyright 1.1.310 permits false negative when matching functools.partial against protocol-based Partial

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#5164]
The following false-negative was introduced in either 1.1.309 or 1.1.310. The following causes 1.1.309 to crash, so I cannot specify the exact version.

```python
from functools import partial
import sys
import types
from typing import TYPE_CHECKING, Any, Callable, Dict, Protocol, Tuple, Type, TypeVar
from typing_extensions import Self

T= TypeVar("T")

class Partial(Protocol[T]):
    __call__: Callable[..., T]

    @property
    def func(self) -> Callable[..., T]:
        ...

    @property
    def args(self) -> Tuple[Any, ...]:
        ...

    @property
    def keywords(self) -> Dict[str, Any]:
        ...

    def __new__(
        cls: Type[Self], __func: Callable[..., T], *args: Any, **kwargs: Any
    ) -> Self:
        ...

    if TYPE_CHECKING and sys.version_info >= (3, 9):  # pragma: no cover

        def __class_getitem__(cls, item: Any) -> types.GenericAlias:
            ...


def f() -> int: ...
def foo(x: Partial[str]): ...

# pyright should report error: Argument of type "partial[int]" cannot be assigned to parameter "x" of type "Partial[str]" 
# in function "foo"
foo(partial(f))   # This becomes a false-negative starting in 1.1.310
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release. Apologies for the regression.

Out of curiosity, how did you notice this? False negatives are typically difficult to detect. Did you happen to have a test case for this?

**Comment by rsokl** [comment]
Yep, I have a test case like:

```python
def f() -> int: ...
def foo(x: Partial[str]): ...

foo(partial(f))  # type: ignore
```

and I was alerted that I had an unnecessary type ignore.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.311, which I just published. It will also be included in a future release of pylance.
